go/service/usersearch.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2019 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package service
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/keybase/client/go/contacts"
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-framed-msgpack-rpc/rpc"
+	"golang.org/x/net/context"
+
+	"golang.org/x/text/unicode/norm"
+)
+
+type UserSearchHandler struct {
+	libkb.Contextified
+	*BaseHandler
+	savedContacts *contacts.SavedContactsStore
+}
+
+func NewUserSearchHandler(xp rpc.Transporter, g *libkb.GlobalContext, pbs *contacts.SavedContactsStore) *UserSearchHandler {
+	handler := &UserSearchHandler{
+		Contextified:  libkb.NewContextified(g),
+		BaseHandler:   NewBaseHandler(g, xp),
+		savedContacts: pbs,
+	}
+	return handler
+}
+
+var _ keybase1.UserSearchInterface = (*UserSearchHandler)(nil)
+
+type rawSearchResults struct {
+	libkb.AppStatusEmbed
+	List []keybase1.APIUserSearchResult `json:"list"`
+}
+
+func doSearchRequest(mctx libkb.MetaContext, arg keybase1.UserSearchArg) (res []keybase1.APIUserSearchResult, err error) {
+	service := arg.Service
+	if service == "keybase" {
+		service = ""
+	}
+	apiArg := libkb.APIArg{
+		Endpoint:    "user/user_search",
+		SessionType: libkb.APISessionTypeNONE,
+		Args: libkb.HTTPArgs{
+			"q":                        libkb.S{Val: arg.Query},
+			"num_wanted":               libkb.I{Val: arg.MaxResults},
+			"service":                  libkb.S{Val: service},
+			"include_services_summary": libkb.B{Val: arg.IncludeServicesSummary},
+		},
+	}
+	var response rawSearchResults
+	err = mctx.G().API.GetDecode(mctx, apiArg, &response)
+	if err != nil {
+		return nil, err
+	}
+	return response.List, nil
+}
+
+func normalizeText(str string) string {
+	return strings.ToLower(string(norm.NFKD.Bytes([]byte(str))))
+}
+
+var splitRxx = regexp.MustCompile(`[-\s!$%^&*()_+|~=` + "`" + `{}\[\]:";'<>?,.\/]+`)
+
+func queryToRegexp(q string) (*regexp.Regexp, error) {
+	parts := splitRxx.Split(q, -1)
+	nonEmptyParts := make([]string, 0, len(parts))
+	for _, p := range parts {
+		if p != "" {
+			nonEmptyParts = append(nonEmptyParts, p)
+		}
+	}
+	rxx, err := regexp.Compile(".*" + strings.Join(nonEmptyParts, ".*") + ".*")
+	if err != nil {
+		return nil, err
+	}
+	rxx.Longest()
+	return rxx, nil
+}
+
+type compiledQuery struct {
+	query string
+	rxx   *regexp.Regexp
+}
+
+func compileQuery(query string) (res compiledQuery, err error) {
+	query = normalizeText(query)
+	rxx, err := queryToRegexp(query)
+	if err != nil {
+		return res, err
+	}
+	res = compiledQuery{
+		query: query,
+		rxx:   rxx,
+	}
+	return res, nil
+}
+
+func (q *compiledQuery) scoreString(str string) (bool, float64) {
```

Comment:
may be a phase 2 thing but we should consider fuzziness. especially people are more likely to misspell people's names (either in query or their addressbook) than usernames.

Reply:
ok, coyne says not necessary and this is fine for now - nvm

Reply:
thanks, good idea though.